### PR TITLE
interfaces: use spec in kmod backend, updated firewall_control, openvswitch_support, ppp

### PR DESCRIPTION
**Pull request by stolowski**
Use spec in the kmod backend. Support new kmod via commonInterface and converted firewall_control, openvswitch_support and ppp interfaces to use it.

Note: depends on https://github.com/snapcore/snapd/pull/2874 (so the diff is actualy smaller once 2874 lands)

### CODE REVIEW

File: interfaces/builtin/common.go [hunk 1]
```diff
@@ -123,3 +126,31 @@ func (iface *commonInterface) ConnectedSlotSnippet(plug *interfaces.Plug, slot *
 func (iface *commonInterface) AutoConnect(*interfaces.Plug, *interfaces.Slot) bool {
 	return !iface.rejectAutoConnectPairs
 }
+
+func (iface *commonInterface) KModConnectedPlug(spec *kmod.Specification, plug *interfaces.Plug, slot *interfaces.Slot) error {
+	for _, m := range iface.connectedPlugKModModules {
+		spec.AddModule(m)
```
Comment (zyga) [inline interfaces/builtin/common.go]: While we don't error in practice in the current code I think we want to check for errors here.

File: interfaces/builtin/common.go [hunk 2]
```diff
@@ -123,3 +126,31 @@ func (iface *commonInterface) ConnectedSlotSnippet(plug *interfaces.Plug, slot *
...
+	}
+	return nil
+}
+
+func (iface *commonInterface) KModConnectedSlot(spec *kmod.Specification, plug *interfaces.Plug, slot *interfaces.Slot) error {
+	for _, m := range iface.connectedSlotKModModules {
+		spec.AddModule(m)
```
Comment (zyga) [inline interfaces/builtin/common.go]: Same as above

File: interfaces/builtin/common.go [hunk 3]
```diff
@@ -123,3 +126,31 @@ func (iface *commonInterface) ConnectedSlotSnippet(plug *interfaces.Plug, slot *
...
+	}
+	return nil
+}
+
+func (iface *commonInterface) KModPermanentPlug(spec *kmod.Specification, plug *interfaces.Plug) error {
+	for _, m := range iface.permanentPlugKModModules {
+		spec.AddModule(m)
```
Comment (zyga) [inline interfaces/builtin/common.go]: and here

File: interfaces/builtin/common.go [hunk 4]
```diff
@@ -123,3 +126,31 @@ func (iface *commonInterface) ConnectedSlotSnippet(plug *interfaces.Plug, slot *
...
+	}
+	return nil
+}
+
+func (iface *commonInterface) KModPermanentSlot(spec *kmod.Specification, slot *interfaces.Slot) error {
+	for _, m := range iface.permanentSlotKModModules {
+		spec.AddModule(m)
```
Comment (zyga) [inline interfaces/builtin/common.go]: and here

File: interfaces/builtin/firewall_control_test.go [hunk 5]
```diff
@@ -87,8 +88,15 @@ func (s *FirewallControlInterfaceSuite) TestUsedSecuritySystems(c *C) {
 	snippet, err = s.iface.ConnectedPlugSnippet(s.plug, s.slot, interfaces.SecuritySecComp)
 	c.Assert(err, IsNil)
 	c.Assert(snippet, Not(IsNil))
-	// connected plugs have a non-nil security snippet for kmod
-	snippet, err = s.iface.ConnectedPlugSnippet(s.plug, s.slot, interfaces.SecurityKMod)
```
Comment (zyga) [inline interfaces/builtin/firewall_control_test.go]: You don't want to copy the definer here or anywhere else. Just create the specification and use the generic specification methods to query the interface. Just call `spec.AddConnectedPlug` and look at what was done to the specification.

File: interfaces/builtin/openvswitch_support.go [hunk 6]
```diff
@@ -30,8 +30,8 @@ openvswitch
 // NewOpenvSwitchSupportInterface returns a new "openvswitch-support" interface.
 func NewOpenvSwitchSupportInterface() interfaces.Interface {
 	return &commonInterface{
-		name:              "openvswitch-support",
-		connectedPlugKMod: openvswitchSupportConnectedPlugKmod,
-		reservedForOS:     true,
+		name: "openvswitch-support",
+		connectedPlugKModModules: []string{"openvswitch"},
```
Comment (zyga) [inline interfaces/builtin/openvswitch_support.go]: Same as earlier, can you please just change that variable above and use the variable here. We can optimise this later but it will make the review more straightforward.

File: interfaces/builtin/openvswitch_support_test.go [hunk 7]
```diff
@@ -79,8 +80,13 @@ func (s *OpenvSwitchSupportInterfaceSuite) TestSanitizeIncorrectInterface(c *C)
 }
 
 func (s *OpenvSwitchSupportInterfaceSuite) TestUsedSecuritySystems(c *C) {
-	// connected plugs have a non-nil security snippet for kmod
-	snippet, err := s.iface.ConnectedPlugSnippet(s.plug, s.slot, interfaces.SecurityKMod)
+	type definer interface {
```
Comment (zyga) [inline interfaces/builtin/openvswitch_support_test.go]: Same comment, don't copy the definer here

File: interfaces/kmod/backend.go [hunk 8]
```diff
@@ -64,18 +64,14 @@ func (b *Backend) Name() interfaces.SecuritySystem {
 func (b *Backend) Setup(snapInfo *snap.Info, confinement interfaces.ConfinementOptions, repo *interfaces.Repository) error {
 	snapName := snapInfo.Name()
 	// Get the snippets that apply to this snap
-	snippets, err := repo.SecuritySnippetsForSnap(snapInfo.Name(), interfaces.SecurityKMod)
+	spec, err := repo.SnapSpecification(b.Name(), snapName)
 	if err != nil {
 		return fmt.Errorf("cannot obtain kmod security snippets for snap %q: %s", snapName, err)
```
Comment (zyga) [inline interfaces/kmod/backend.go]: You can tweak this error message as we're not using snippets anymore. Look at the mount backend and use similar wording.

File: interfaces/kmod/backend.go [hunk 9]
```diff
@@ -103,54 +99,27 @@ func (b *Backend) Remove(snapName string) error {
...
-				if len(l) > 0 && l[0] != '#' {
-					modules = append(modules, string(l))
-				}
-			}
-		}
+func deriveContent(spec *Specification, snapInfo *snap.Info) (map[string]*osutil.FileState, []string) {
+	content := make(map[string]*osutil.FileState)
+	var modules []string
+	for k := range spec.Modules {
+		modules = append(modules, k)
```
Comment (zyga) [inline interfaces/kmod/backend.go]: Can you please sort them afterwards. This will be important soon.

File: interfaces/kmod/backend.go [hunk 10]
```diff
@@ -103,54 +99,27 @@ func (b *Backend) Remove(snapName string) error {
...
-			for _, line := range bytes.Split(snippet, []byte{'\n'}) {
-				l := bytes.TrimSpace(line)
-				// ignore empty lines and comments
-				if len(l) > 0 && l[0] != '#' {
-					modules = append(modules, string(l))
-				}
-			}
-		}
+func deriveContent(spec *Specification, snapInfo *snap.Info) (map[string]*osutil.FileState, []string) {
+	content := make(map[string]*osutil.FileState)
```
Comment (zyga) [inline interfaces/kmod/backend.go]: This can be all short-circuited if there are no modules in the spec. You don't have to create the map or do anything else when that is the case. Most interfaces will not use this backend so it is worth doing.

File: interfaces/kmod/backend_test.go [hunk 11]
```diff
@@ -117,11 +98,9 @@ func (s *backendSuite) TestInstallingSnapCreatesModulesConf(c *C) {
 
 func (s *backendSuite) TestRemovingSnapRemovesModulesConf(c *C) {
 	// NOTE: Hand out a permanent snippet so that .conf file is generated.
-	s.Iface.PermanentSlotSnippetCallback = func(slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
-		if securitySystem == interfaces.SecurityKMod {
-			return []byte("module1\nmodule2"), nil
-		}
-		return nil, nil
+	s.Iface.KModPermanentSlotCallback = func(spec *kmod.Specification, slot *interfaces.Slot) error {
+		spec.AddModule("module1")
```
Comment (zyga) [inline interfaces/kmod/backend_test.go]: The original code added module1 and module2 here, why did you change this?

File: interfaces/kmod/backend_test.go [hunk 12]
```diff
@@ -137,12 +116,11 @@ func (s *backendSuite) TestRemovingSnapRemovesModulesConf(c *C) {
 
 func (s *backendSuite) TestSecurityIsStable(c *C) {
 	// NOTE: Hand out a permanent snippet so that .conf file is generated.
-	s.Iface.PermanentSlotSnippetCallback = func(slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
-		if securitySystem == interfaces.SecurityKMod {
-			return []byte("module1\nmodule2"), nil
-		}
-		return nil, nil
+	s.Iface.KModPermanentSlotCallback = func(spec *kmod.Specification, slot *interfaces.Slot) error {
+		spec.AddModule("module1")
```
Comment (zyga) [inline interfaces/kmod/backend_test.go]: Ditto

File: interfaces/kmod/backend_test.go (hunk 11 above)
Comment (stolowski) [inline interfaces/kmod/backend_test.go]: This callback is irrelevant for this test case, it's there just to have a non-empty list of modules, but what we're really testing is that the generated file is removed when snap is uninstalled.

File: interfaces/kmod/backend_test.go (hunk 12 above)
Comment (stolowski) [inline interfaces/kmod/backend_test.go]: It's similar to the earlier case, although in this case I agree it makes sense to have 2 modules returned as I'm testing if security is stable, multiple modules make a batter test case.

File: interfaces/kmod/backend.go (hunk 9 above)
Comment (stolowski) [inline interfaces/kmod/backend.go]: What do you mean? I sort them immediately after this loop, do you mean to sort them elsewhere?

File: interfaces/kmod/backend.go (hunk 8 above)
Comment (stolowski) [inline interfaces/kmod/backend.go]: From the mount backend:
"return fmt.Errorf("cannot obtain mount security snippets for snap %q: %s", snapName, err)"
;)
How about: "cannot obtain kmod specification for snap %q: %s"... ?

Comment (zyga) [inline interfaces/kmod/backend.go]: Ah, I'm sorry for the useless advice. I changed this in my apparmor branch locally. Your suggestion reads good. Lets use it.

File: interfaces/kmod/backend.go (hunk 9 above)
Comment (zyga) [inline interfaces/kmod/backend.go]: Ah, sorry I missed that somehow.

File: interfaces/builtin/ppp.go [hunk 13]
```diff
@@ -77,6 +74,10 @@ func (iface *PppInterface) ConnectedSlotSnippet(plug *interfaces.Plug, slot *int
 	return nil, nil
 }
 
+func (iface *PppInterface) KModConnectedPlug(spec *kmod.Specification, plug *interfaces.Plug, slot *interfaces.Slot) error {
```
Comment (zyga) [inline interfaces/builtin/ppp.go]: Nice, looking forward to not having to copy-paste empty functions when this work is over :-)

File: interfaces/kmod/backend.go [hunk 14]
```diff
@@ -64,18 +64,17 @@ func (b *Backend) Name() interfaces.SecuritySystem {
...
+		return fmt.Errorf("cannot obtain kmod specification for snap %q: %s", snapName, err)
 	}
 
-	// Get the files that this snap should have
-	glob := interfaces.SecurityTagGlob(snapName)
-	content, modules, err := b.combineSnippets(snapInfo, snippets)
-	if err != nil {
-		return fmt.Errorf("cannot obtain expected security files for snap %q: %s", snapName, err)
+	if len(spec.(*Specification).Modules) == 0 {
+		return nil
```
Comment (zyga) [inline interfaces/kmod/backend.go]: No, this part is wrong. This should go to the `deriveContent` call. You absolutely have to call `EnsureDirState` below as that is what really applies the changes to the system. You also need to act on the things it returns.

If you move this to to `deriveContent` then you save useless construction of one map. That what what I was suggesting earlier (since most interfaces don't need kmod)

File: interfaces/kmod/backend.go [hunk 15]
```diff
@@ -103,54 +102,27 @@ func (b *Backend) Remove(snapName string) error {
...
-			}
-		}
+func deriveContent(spec *Specification, snapInfo *snap.Info) (map[string]*osutil.FileState, []string) {
+	content := make(map[string]*osutil.FileState)
+	var modules []string
+	for k := range spec.Modules {
+		modules = append(modules, k)
 	}
-
 	sort.Strings(modules)
```
Comment (zyga) [inline interfaces/kmod/backend.go]: Sorry, I didn't see this somehow  +1

File: interfaces/kmod/backend.go [hunk 16]
```diff
@@ -103,54 +102,27 @@ func (b *Backend) Remove(snapName string) error {
...
 
-		content[fmt.Sprintf("%s.conf", snap.SecurityTag(snapInfo.Name()))] = &osutil.FileState{
-			Content: buffer.Bytes(),
-			Mode:    0644,
-		}
+	var buffer bytes.Buffer
+	buffer.WriteString("# This file is automatically generated.\n")
+	for _, module := range modules {
+		buffer.WriteString(module)
+		buffer.WriteByte('\n')
```
Comment (zyga) [inline interfaces/kmod/backend.go]: Nitpick: `WriteRune('\n')`

File: interfaces/kmod/backend.go (hunk 14 above)
Comment (stolowski) [inline interfaces/kmod/backend.go]: Ok, fixed.

File: interfaces/kmod/spec.go [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+// holds internal state that is used by the kmod backend during the interface
+// setup process.
+type Specification struct {
+	Modules map[string]bool
+}
+
+// AddModule adds a kernel module, trimming spaces and ignoring duplicated modules.
+func (spec *Specification) AddModule(module string) error {
+	m := strings.TrimSpace(module)
+	if len(m) > 0 {
```
Comment (mvo5) [inline interfaces/kmod/spec.go]: You could just write `if len(m) == 0 { return nil }` here (or given that its a string even better: `if m == "" { return nil }`. This is more common in go (return early and have the rest of the function in the normal function scope instead of on a if).

Comment (stolowski) [inline interfaces/kmod/spec.go]: Ok, done

**Comment by zyga** [comment]
`interfaces/kmod/spec.go:43: comparison of function Modules == nil is always false`

```
interfaces/kmod/backend.go:103: invalid argument spec.Modules (type func() map[string]bool) for len
interfaces/kmod/backend.go:108: cannot range over spec.Modules (type func() map[string]bool)
interfaces/kmod/spec.go:44: cannot assign to spec.Modules
interfaces/kmod/spec.go:44: cannot use make(map[string]bool) (type map[string]bool) as type func() map[string]bool in assignment
interfaces/kmod/spec.go:46: invalid operation: spec.Modules[m] (type func() map[string]bool does not support indexing)
```
